ingestion-beam/src/main/java/com/mozilla/telemetry/util/KeyStore.java

```diff
@@ -0,0 +1,122 @@
+import java.security.PrivateKey;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.io.IOUtils;
+import org.everit.json.schema.Schema;
+import org.everit.json.schema.loader.SchemaLoader;
+import org.jose4j.jwk.PublicJsonWebKey;
+import org.jose4j.lang.JoseException;
+
+/**
+ * Store Private Keys derived from JSON Web Keys (JWK) decrypted via the GCP Key
+ * Management Service (KMS).
+ */
+public class KeyStore {
+
+  /**
+   * Fetch a key from the store.
+   *
+   * @param path fully qualified path to the key
+   * @return private key for decryption
+   */
+  public PrivateKey getKey(String path) {
+    ensureKeysLoaded();
+    PrivateKey key = keys.get(path);
+    return key;
+  }
+
+  private final String metadataLocation;
+  private final boolean kmsEnabled;
+  private transient Map<String, PrivateKey> keys;
+
+  public static KeyStore of(String metadataLocation, boolean kmsEnabled) {
+    return new KeyStore(metadataLocation, kmsEnabled);
+  }
+
+  private KeyStore(String metadataLocation, boolean kmsEnabled) {
+    this.metadataLocation = metadataLocation;
+    this.kmsEnabled = kmsEnabled;
+  }
+
+  @VisibleForTesting
+  int numLoadedKeys() {
+    ensureKeysLoaded();
+    return keys.size();
+  }
+
+  private void loadAllKeys() throws IOException, ValidationException {
+    final Map<String, PrivateKey> tempKeys = new HashMap<>();
+
+    Schema schema;
+    try (InputStream inputStream = Resources.getResource("keystore-metadata.schema.json")
+        .openStream()) {
+      schema = SchemaLoader.load(Json.readJsonObject(inputStream));
+    } catch (IOException e) {
+      throw new IOException("Exception thrown while reading metadata file", e);
+    }
+
+    // required to validate Jackson objects
+    JsonValidator validator = new JsonValidator();
+
+    ArrayNode metadata;
+    try (InputStream inputStream = BeamFileInputStream.open(this.metadataLocation)) {
+      byte[] data = IOUtils.toByteArray(inputStream);
+      metadata = Json.readArrayNode(data);
+      validator.validate(schema, metadata);
+    } catch (IOException e) {
+      throw new IOException("Exception thrown while reading keystore metadata schema.", e);
+    }
+
+    for (JsonNode element : metadata) {
+      String privateKeyId = element.get("private_key_id").textValue();
+      String privateKeyUri = element.get("private_key_uri").textValue();
+      String kmsResourceId = element.get("kms_resource_id").textValue();
+
+      try (InputStream inputStream = BeamFileInputStream.open(privateKeyUri)) {
+        byte[] keyData = IOUtils.toByteArray(inputStream);
+
+        PublicJsonWebKey key;
+        if (kmsEnabled) {
+          try (KeyManagementServiceClient client = KeyManagementServiceClient.create()) {
+            DecryptResponse response = client.decrypt(kmsResourceId, ByteString.copyFrom(keyData));
+            key = PublicJsonWebKey.Factory
+                .newPublicJwk(new String(response.getPlaintext().toByteArray()));
```

Comment:
We should always specify `StandardCharsets.UTF_8` when converting from string to bytes or vice versa.